### PR TITLE
Activity feed broadcast csharp

### PR DESCRIPTION
**Pull request by NikitaJain-MSFT**

### CODE REVIEW

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                    ViewBag.TaskList = taskList;
+                }
+                else
+                {
+                    currentTaskList.Add(taskInfo);
+                    _taskList.AddOrUpdate("taskList", currentTaskList, (key, newValue) => currentTaskList);
+                    ViewBag.TaskList = currentTaskList;
+                }
+
+                var graphClient = SimpleGraphClient.GetGraphClient(taskInfo.access_token);
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:98]: Do  we need separate variables for graphClient? if yes, we can add some comment here.

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                var graphClientApp = SimpleGraphClient.GetGraphClientforApp(_configuration["AzureAd:MicrosoftAppId"], _configuration["AzureAd:MicrosoftAppPassword"], _configuration["AzureAd:TenantId"]);
+                var usersList = await graphClient.Users
+                                 .Request()
+                                    .GetAsync();
+
+                var installedAppsForCurrentUser = await graphClient.Users[taskInfo.userId].Teamwork.InstalledApps
+                                   .Request()
+                                   .Expand("teamsAppDefinition")
+                                   .GetAsync();
+                var appId = installedAppsForCurrentUser.Where(id => id.TeamsAppDefinition.DisplayName == "Activity feed broadcast").Select(x => x.TeamsAppDefinition.TeamsAppId);
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:108]: We should provide this information about the Display name being used for the condition.

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                                var exponentialBackoff = Math.Pow(2, retryCount);
+                                var delayInSeconds = exponentialBackoff * 10000;
+                                delay = TimeSpan.FromMilliseconds(delayInSeconds);
+                            }
+
+                            return delay;
+                        },
+                        onRetryAsync: async (response, timespan, retryCount, context) =>
+                        {
+                        }
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:196]: We can remove this part as it is empty.

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/SimpleGraphClient.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using Microsoft.Graph;
+using Microsoft.Identity.Client;
+
+namespace ActivityFeedBroadcast
+{
+    public class SimpleGraphClient
+    {
+        public static GraphServiceClient GetGraphClient(string accessToken)
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/SimpleGraphClient.cs:11]: Some comments for method will be good to have.

File: samples/graph-activity-feed-broadcast/csharp/README.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+18. Setup and run the bot from Visual Studio: 
+   Modify the `appsettings.json` and fill in the following details:
+   - `MicrosoftAppId` - Generated from Step 3 (Application (client) ID)is the application app id
+   - `TenantId` - Generated from Step 3(Directory (tenant) ID) is the tenant id
+   - `MicrosoftAppPassword` - Generated from Step 14, also referred to as Client secret
+   - `{Base_URL}` - Your application's base url. E.g. https://12345.ngrok.io if you are using ngrok.
+   - Press `F5` to run the project
+	 
+19. Modify the `manifest.json` in the `/AppPackage` folder and replace the following details:
+   - `{{Microsoft-App-Id}}` with Application id generated from Step 3
```
Comment (AbhijeetThakur-msft) [inline samples/graph-activity-feed-broadcast/csharp/README.md:124]: cross check if the keys are used in same manner in manifest.json file e.g. spelling for Microsoft-App-Id

**Review by AbhijeetThakur-msft — CHANGES_REQUESTED** [review]

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 1 above)
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:99]: Added comments

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 2 above)
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:113]: added comment and added to readme

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs (hunk 3 above)
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:201]: onRetryAsync is mandatory parameter.

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                    var client = new HttpClient();
+                    var url = "https://teams.microsoft.com/l/entity/" + appId.ToList()[0] + "/broadcast?context={\"subEntityId\":\"" + taskInfo.taskId + "\"}";
+                    var postData = new
+                    {
+                        topic = new
+                        {
+                            source = "text",
+                            value = $"{taskInfo.title}",
+                            webUrl = url
+                        },
```
[the remark this answers is not in the document]
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs]: updated with sdk classes

File: samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+                            }
+                        };
+
+                        await graphClient.Users[users.Id].Teamwork.InstalledApps
+                                            .Request()
+                                            .AddAsync(userScopeTeamsAppInstallation);
+
+                        client.DefaultRequestHeaders.Authorization =
+                           new AuthenticationHeaderValue("Bearer", taskInfo.access_token);
+                        await client.PostAsync($"https://graph.microsoft.com/v1.0/users/{users.Id}/teamwork/sendActivityNotification", data);
```
[the remark this answers is not in the document]
Comment (NikitaJain-MSFT) [reply samples/graph-activity-feed-broadcast/csharp/ActivityFeedBroadcast/Controllers/HomeController.cs:167]: For send activity notification call graph SDk method does not return any value and we require status code of graph call.

**Review by AbhijeetThakur-msft — APPROVED** [review]

**Review by v-chetsh — APPROVED** [review]
Changes looks fine, approving it.

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**